pkg/eval/unix/umask.go

```diff
@@ -0,0 +1,94 @@
+// +build !windows,!plan9
+
+package unix
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+	"syscall"
+
+	"github.com/elves/elvish/pkg/eval"
+	"github.com/elves/elvish/pkg/eval/vars"
+)
+
+//elvdoc:var umask
+//
+// The file system creation mask. When used in a string context the output is
+// its Elvish octal representation; e.g. 0o027. This makes it possible to use
+// it in any context that expects a `$number`.
+//
+// When assigning a new value a string is implicitly treated as an octal
+// number. If that fails the usual rules for interpreting
+// [numbers](./language.html#number) are used. This means you can do both
+// `unix:umask = 027` and `unix:umask = 0o27`. You can also assign to it a
+// `float64` data type that has no fractional component. The legal values are
+// in the range [0 .. 0o777].
+//
+// You can do a temporary assignment to affect a single command; e.g.
+// `umask=077 touch a_file`. After the command completes the old umask will be
+// restored. **Warning**: Since the umask applies to the entire process, not
+// individual threads, changing it temporarily in this manner is dangerous if
+// you are doing anything in parallel. Such as via the
+// [`peach`](ref/builtin.html#peach) command.
+//
+// This variable only has meaning on UNIX like operating systems and does not
+// exist on other platforms. Do not use this in elvish scripts meant to work
+// on non-UNIX compatible systems.
+
+// UmaskVariable is a variable whose value always reflects the current file
+// creation permission mask. Setting it changes the current file creation
+// permission mask for the process (not an individual thread).
+type UmaskVariable struct {
+	ev *eval.Evaler
+}
+
+var _ vars.Var = UmaskVariable{}
+
+// Get returns the current file creation umask as a string.
+func (UmaskVariable) Get() interface{} {
+	// Note: The seemingly redundant syscall is because the syscall.Umask() API
+	// doesn't allow querying the current value without changing it. So ensure
+	// we reinstate the curent value.
+	umask := syscall.Umask(0)
+	syscall.Umask(umask)
+	return fmt.Sprintf("0o%03o", umask)
+}
+
+// Set changes the current file creation umask. It can be called with a string
+// (the usual case) or a float64.
+func (UmaskVariable) Set(v interface{}) error {
+	var umask int
+
+	switch v := v.(type) {
+	case string:
+		i, err := strconv.ParseInt(v, 8, 0)
+		if err != nil {
+			i, err = strconv.ParseInt(v, 0, 0)
+			if err != nil {
+				return errors.New("umask value not a valid number")
+			}
+		}
+		umask = int(i)
+	case float64:
+		int_part, frac_part := math.Modf(v)
```

Comment:
Do not use underscores in variable names.

Reply:
Breaking that habit, instilled by3+ decades of doing that in nearly all the C code I've ever authored or touched, is going to take a few applications of the old clue-by-four :smile: